File: THCS/2023_2024/HaNoi_2324/Bai4/ROBOT.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <algorithm>
using namespace std;

// Hàm giải bài toán robot
vector<int> solve_robot_problem(int N, int M, int Q, int K, const vector<vector<int>>& A, const vector<int>& queries) {
    // Tạo một mảng dp để lưu số lượng ô thỏa mãn điều kiện Ai,j % K = X cho mỗi X
    vector<vector<vector<int>>> dp(N, vector<vector<int>>(M, vector<int>(K, 0)));
    
    // Khởi tạo cho ô (1, 1)
    dp[0][0][A[0][0] % K] = 1;
    
    // Tính toán cho toàn bộ bảng A
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < M; ++j) {
            if (i == 0 && j == 0) continue; // Bỏ qua ô khởi đầu
            
            int current_mod = A[i][j] % K;
            
            // Lấy giá trị lớn nhất từ các ô (i-1, j), (i, j-1), (i-1, j-1)
            for (int mod = 0; mod < K; ++mod) {
                int max_prev = 0;
                if (i > 0) {
                    max_prev = max(max_prev, dp[i-1][j][mod]);
                }
                if (j > 0) {
                    max_prev = max(max_prev, dp[i][j-1][mod]);
                }
                if (i > 0 && j > 0) {
                    max_prev = max(max_prev, dp[i-1][j-1][mod]);
                }
                
                dp[i][j][mod] = max_prev;
            }
            
            // Cộng thêm 1 cho giá trị mod hiện tại tại ô (i,j)
            dp[i][j][current_mod] += 1;
        }
    }
    
    // Trả lời các truy vấn
    vector<int> result;
    for (int X : queries) {
        result.push_back(dp[N-1][M-1][X]);
    }
    
    return result;
}

// Hàm đọc dữ liệu từ tệp
void read_input(const string& filename, int& N, int& M, int& Q, int& K, vector<vector<int>>& A, vector<int>& queries) {
    ifstream infile(filename);
    infile >> N >> M >> Q >> K;
    
    A.resize(N, vector<int>(M));
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < M; ++j) {
            infile >> A[i][j];
        }
    }
    
    queries.resize(Q);
    for (int i = 0; i < Q; ++i) {
        infile >> queries[i];
    }
    
    infile.close();
}

// Hàm ghi kết quả ra tệp
void write_output(const string& filename, const vector<int>& result) {
    ofstream outfile(filename);
    for (int r : result) {
        outfile << r << endl;
    }
    outfile.close();
}

int main() {
    int N, M, Q, K;
    vector<vector<int>> A;
    vector<int> queries;
    
    // Đọc dữ liệu từ tệp ROBOT.INP
    read_input("ROBOT.INP", N, M, Q, K, A, queries);
    
    // Giải bài toán
    vector<int> result = solve_robot_problem(N, M, Q, K, A, queries);
    
    // Ghi kết quả ra tệp ROBOT.OUT
    write_output("ROBOT.OUT", result);
    
    return 0;
}

File: THCS/2023_2024/HaNoi_2324/Bai4/ROBOT.py
def solve_robot_problem(N, M, Q, K, A, queries):
    # Tạo một mảng để lưu số lượng ô thỏa mãn điều kiện Ai,j % K = X cho mỗi X
    dp = [[{i: 0 for i in range(K)} for _ in range(M)] for _ in range(N)]
    
    # Khởi tạo cho ô (1, 1)
    dp[0][0][A[0][0] % K] = 1
    
    # Tính toán cho toàn bộ bảng A
    for i in range(N):
        for j in range(M):
            if i == 0 and j == 0:
                continue  # Bỏ qua ô khởi đầu đã tính trước
            
            current_mod = A[i][j] % K
            
            # Lấy giá trị lớn nhất từ các ô (i-1, j), (i, j-1), (i-1, j-1)
            for mod in range(K):
                max_prev = 0
                if i > 0:
                    max_prev = max(max_prev, dp[i-1][j][mod])
                if j > 0:
                    max_prev = max(max_prev, dp[i][j-1][mod])
                if i > 0 and j > 0:
                    max_prev = max(max_prev, dp[i-1][j-1][mod])
                
                dp[i][j][mod] = max_prev
            
            # Cộng thêm 1 cho giá trị mod hiện tại tại ô (i,j)
            dp[i][j][current_mod] += 1
    
    # Trả lời các truy vấn
    result = []
    for X in queries:
        result.append(dp[N-1][M-1][X])
    
    return result


# Đọc dữ liệu từ tệp
def read_input(filename):
    with open(filename, 'r') as f:
        N, M, Q, K = map(int, f.readline().split())
        A = [list(map(int, f.readline().split())) for _ in range(N)]
        queries = [int(f.readline().strip()) for _ in range(Q)]
    return N, M, Q, K, A, queries

# Ghi kết quả ra tệp
def write_output(filename, result):
    with open(filename, 'w') as f:
        for r in result:
            f.write(str(r) + '\n')

# Chương trình chính
if __name__ == "__main__":
    N, M, Q, K, A, queries = read_input('ROBOT.INP')
    result = solve_robot_problem(N, M, Q, K, A, queries)
    write_output('ROBOT.OUT', result)
